File: modules/agent/src/models/Tag.py
from sqlalchemy import Column, BigInteger, String, Text, DateTime, ForeignKey, func
from sqlalchemy.orm import relationship
from models import Base


class EntityTag(Base):
    """Polymorphic junction for tagging any entity."""

    __tablename__ = "Entity_Tag"

    tag_id = Column(BigInteger, ForeignKey("Tag.id", ondelete="CASCADE"), primary_key=True)
    entity_type = Column(Text, primary_key=True)
    entity_id = Column(BigInteger, primary_key=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    # Relationships
    tag = relationship("Tag", back_populates="entity_links")


class Tag(Base):
    """Unique tags for categorizing any entity."""

    __tablename__ = "Tag"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    name = Column(String(100), nullable=False, unique=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    # Relationships
    entity_links = relationship("EntityTag", back_populates="tag")

File: modules/agent-go/internal/models/Tag.go
package models

import "time"

type Tag struct {
	ID        int64     `gorm:"primaryKey;autoIncrement" json:"id"`
	TenantID  *int64    `gorm:"index" json:"tenant_id"`
	Name      string    `gorm:"size:100;not null" json:"name"` // unique per tenant
	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`
}

func (Tag) TableName() string {
	return "Tag"
}

type EntityTag struct {
	TagID      int64     `gorm:"primaryKey" json:"tag_id"`
	EntityType string    `gorm:"primaryKey;type:text" json:"entity_type"`
	EntityID   int64     `gorm:"primaryKey" json:"entity_id"`
	CreatedAt  time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt  time.Time `gorm:"autoUpdateTime" json:"updated_at"`
}

func (EntityTag) TableName() string {
	return "Entity_Tag"
}
